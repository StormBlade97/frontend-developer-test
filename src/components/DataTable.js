import * as React from 'react';
import Table from '@material-ui/core/Table';
import TableBody from '@material-ui/core/TableBody';
import TableCell from '@material-ui/core/TableCell';
import TableContainer from '@material-ui/core/TableContainer';
import TableHead from '@material-ui/core/TableHead';
import TableRow from '@material-ui/core/TableRow';
import TableSortLabel from '@material-ui/core/TableSortLabel';
import Paper from '@material-ui/core/Paper';

const DataTable = ({ columns, rows, tableController, sortBy, sortDirection, toggleSort }) => {
    const hasData = rows && rows.length > 0;
    
    return <TableContainer component={Paper}>
        <Table aria-label="simple table">
            <TableHead>
                <TableRow>
                    {
                        columns.map(header =>
                            <TableCell component="th" key={header}>
                                <TableSortLabel
                                    active={sortBy === header}
                                    direction={sortDirection}
                                    onClick={toggleSort}
                                >
                                    {header}
                                </TableSortLabel>
                             </TableCell>)
                    }
                </TableRow>
            </TableHead>
            <TableBody>
                {hasData && rows.map((row, idx) => (
                    <TableRow key={idx}>
                        <TableCell>{row.date}</TableCell>
                        <TableCell>
                            {row.userid}
                        </TableCell>
                        <TableCell>{row.oldVal}</TableCell>
                        <TableCell>{row.newVal}</TableCell>
                    </TableRow>
                ))}
                <TableRow>
                    <TableCell align="center" colSpan={columns.length}>
                        {
                            tableController
                        }
                    </TableCell>
                </TableRow>
            </TableBody>
        </Table>
    </TableContainer>;
}

export default DataTable;