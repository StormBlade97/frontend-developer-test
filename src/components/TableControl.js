import React from 'react';
import { FETCH_STATE } from '../const';
import Box from '@material-ui/core/Box';
import Button from '@material-ui/core/Button';
import Typography from '@material-ui/core/Typography';
import CircularProgress from '@material-ui/core/CircularProgress';


const TableControl = ({ fetchState, fetchData }) => {
    const isError = fetchState === FETCH_STATE.ERROR;
    const isLoading = fetchState === FETCH_STATE.PENDING;
    const buttonLabel = isError ? "Retry" : "Load more";

    return <Box p={1}>
        {isError &&
            <Box m={1}>
                <Typography color="error">We had problems fetching your data. Please try again</Typography>
            </Box>
        }
        {
            isLoading ? <CircularProgress ></CircularProgress> :
                <Button variant="contained" color="primary" onClick={fetchData}>
                    {buttonLabel}
                </Button>
        }
    </Box>
}

export default TableControl;