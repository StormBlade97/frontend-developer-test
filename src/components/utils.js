import { compareAsc, compareDesc } from 'date-fns';

export const sortByDate = sortDir => (d1, d2) => {
    const sortFn = sortDir === 'asc' ? compareAsc : compareDesc;
    return sortFn(d1.date, d2.date);
}
export const mapToTableData = (data) => {
    return data.map((entry, idx) => {
        return {
            id: idx,
            dedupId: `${entry.timestamp}${entry.id}`,
            date: entry.timestamp,
            userid: entry.id,
            oldVal: entry.diff[0].oldValue,
            newVal: entry.diff[0].newValue
        }
    })
}
    
export const uniqify = (arr1, arr2) => {
    return [...new Set([...arr1, ...arr2])];
}

export const uniqueBy = (arr, key) => {
    const seen = [];
    const result = [];

    arr.forEach(elem => {
        if (!seen.includes(elem[key])) {
            seen.push(elem[key]);
            result.push(elem);
        }
    })
    return result;
}