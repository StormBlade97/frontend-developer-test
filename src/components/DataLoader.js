import React from 'react';
import { FETCH_STATE } from '../const';

import Box from '@material-ui/core/Box';
import DataTable from './DataTable';
import TableControl from './TableControl';
import { format } from 'date-fns';
import { sortByDate, mapToTableData, uniqueBy } from './utils';

class DataLoader extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            data: [],
            sortDir: 'asc',
            errorMessage: '',
            fetchState: FETCH_STATE.INITIAL
        }
    }
    
    handleFetchData = async () => {
        try {
            this.setState({ ...this.state, fetchState: FETCH_STATE.PENDING });
            const result = await this.props.api();
            
            this.setState({
                ...this.state, fetchState: FETCH_STATE.SUCCESSFUL,
                data: uniqueBy([...this.state.data, ...result.data], 'timestamp')
            });
        }
        catch (error) {
            this.setState({ ...this.state, fetchState: FETCH_STATE.ERROR, errorMessage: error.message});
        }
    }

    toggleSortDir = async () => {
        this.setState({
            ...this.state,
            sortDir: this.state.sortDir === 'asc' ? 'desc' : 'asc'
        });
    }

    componentDidMount() {
        this.handleFetchData();
    }

    render() {
        const { sortDir, data, errorMessage, fetchState } = this.state;
        const sortFn = sortByDate(this.state.sortDir);
        const rows = mapToTableData(data)
            .sort(sortFn)
            .map(({ date, ...rest }) => ({ ...rest, date: format(date, 'yyyy-MM-dd') }));
    
        return (
            <Box data-test="app-box" m={2}>
                <DataTable
                    errorMessage={errorMessage}
                    fetchState={fetchState}
                    columns={['Date', 'User id', 'Old value', 'New value']}
                    sortBy={'Date'}
                    sortDirection={sortDir}
                    toggleSort={this.toggleSortDir}
                    rows={rows}
                    tableController={
                        <TableControl
                            data-test='table-controller'
                            fetchState={fetchState}
                            fetchData={this.handleFetchData} />
                    }
                />
            </Box>
        );
    }
}
export default DataLoader;