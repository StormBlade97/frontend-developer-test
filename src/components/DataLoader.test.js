import React from 'react';
import { shallow, mount } from 'enzyme';
import DataLoader from './DataLoader';
import DataTable from './DataTable';
import TableControl from './TableControl';
import { act } from 'react-dom/test-utils';
import { usersDiff } from '../lib/api/data';
import * as utilsModule from './utils';

describe("DataLoader", () => {
    describe("data fetching", () => {
        it("has correct default state", async () => {
            const apiFn = jest.fn(() => Promise.resolve([{}]));
            const wrapper = shallow(<DataLoader api={apiFn}></DataLoader>);

            expect(wrapper.state()).toMatchSnapshot()
        });
        it("fetches and accumulates unique data entries", async () => {
            const [dataSet1, dataSet2, ...unused] = usersDiff;
            const apiFn = jest.fn()
                .mockReturnValueOnce(Promise.resolve({ data: [dataSet1] }))
                .mockReturnValueOnce(Promise.resolve({ data: [dataSet2] }))
                .mockReturnValueOnce(Promise.resolve({ data: [dataSet2] }));

            const wrapper = shallow(<DataLoader api={apiFn}></DataLoader>);
            const fetchData = wrapper.instance().handleFetchData;
                
            await fetchData();
            await fetchData();

            expect(apiFn.mock.calls.length).toBe(3);
            expect(wrapper.state()).toMatchSnapshot();
        });

        it("sorts newly fetched data", () => {
            const [dataSet1, dataSet2, ...unused] = usersDiff;
            const sortFnSpy = jest.spyOn(utilsModule, 'sortByDate');

            const apiFn = jest.fn()
                .mockReturnValueOnce(Promise.resolve([dataSet1]))
                .mockReturnValueOnce(Promise.resolve([dataSet2]));

            const wrapper = shallow(<DataLoader api={apiFn}></DataLoader>);
            const fetchData = wrapper.find(DataTable).props().tableController.props.fetchData;
                
            fetchData();
            fetchData();

            expect(sortFnSpy.mock.calls).toMatchSnapshot();
        })
    })
    describe("effects", () => {
        it("calls API on mount", () => {
            const apiFn = jest.fn()
            shallow(<DataLoader api={apiFn}></DataLoader>);

            expect(apiFn).toBeCalledTimes(1);
        })
    })
    describe("renders", () => {
        it("render DataTable", () => {
            const apiFn = jest.fn()
            expect(shallow(<DataLoader api={apiFn}></DataLoader>).debug()).toMatchSnapshot();
        });
        it("provides correct props to DataTable", () => {
            const apiFn = jest.fn()
            const wrapper = shallow(<DataLoader api={apiFn}></DataLoader>);
            expect(wrapper.find(DataTable).props()).toMatchSnapshot();
        });
        it("provides correct props to TableController", () => {});
    })
})