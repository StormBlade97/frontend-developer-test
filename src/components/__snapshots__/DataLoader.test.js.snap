// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataLoader data fetching fetches and accumulates unique data entries 1`] = `
Object {
  "data": Array [
    Object {
      "diff": Array [
        Object {
          "field": "name",
          "newValue": "Bruce",
          "oldValue": "John",
        },
      ],
      "id": "e28d290a-a2f2-48c2-9001-ff43884e271b",
      "timestamp": 1581631200000,
    },
    Object {
      "diff": Array [
        Object {
          "field": "name",
          "newValue": "Nick",
          "oldValue": "Bruce",
        },
      ],
      "id": "8bd0166f-f0c6-48fd-9fcd-a17e76eb1e92",
      "timestamp": 1581717600000,
    },
  ],
  "errorMessage": "",
  "fetchState": "successful",
  "sortDir": "asc",
}
`;

exports[`DataLoader data fetching has correct default state 1`] = `
Object {
  "data": Array [],
  "errorMessage": "",
  "fetchState": "pending",
  "sortDir": "asc",
}
`;

exports[`DataLoader data fetching sorts newly fetched data 1`] = `
Array [
  Array [
    "asc",
  ],
  Array [
    "asc",
  ],
  Array [
    "asc",
  ],
  Array [
    "asc",
  ],
]
`;

exports[`DataLoader renders provides correct props to DataTable 1`] = `
Object {
  "columns": Array [
    "Date",
    "User id",
    "Old value",
    "New value",
  ],
  "errorMessage": "",
  "fetchState": "pending",
  "rows": Array [],
  "sortBy": "Date",
  "sortDirection": "asc",
  "tableController": <TableControl
    data-test="table-controller"
    fetchData={[Function]}
    fetchState="pending"
  />,
  "toggleSort": [Function],
}
`;

exports[`DataLoader renders render DataTable 1`] = `
"<Styled(MuiBox) data-test=\\"app-box\\" m={2}>
  <DataTable errorMessage=\\"\\" fetchState=\\"pending\\" columns={{...}} sortBy=\\"Date\\" sortDirection=\\"asc\\" toggleSort={[Function]} rows={{...}} tableController={{...}} />
</Styled(MuiBox)>"
`;
